### ISSUE TITLE
Deadlock when resolving dependencies with Eclipse plugin

### ISSUE THREAD

**Issue by pln-rec** [gradle/gradle#3579]
When using `gradlew eclipseClasspath`, Gradle gets stuck after some seconds.

### Expected Behavior
Gradle does not get stuck. The problem was not present with Gradle 2.9 which we used until now.

### Current Behavior
Gradle gets stuck and does not advance in any way (observed for hours). On the command line this looks like:
```
<-------------> 1% EXECUTING [50m 27s]
> :sub_project1:eclipseClasspath
> :sub_project2:eclipseClasspath > Resolve dependencies :sub_project2:provided
> :sub_project3:eclipseClasspath > Resolve dependencies :sub_project3:testCompileClasspath
> :sub_project4:eclipseClasspath > Resolve dependencies :sub_project4:provided
> :sub_project5:eclipseClasspath
> :sub_project6:eclipseClasspath > Resolve dependencies :sub_project6:compileClasspath
> :sub_project7:eclipseClasspath
> :sub_project8:eclipseClasspath > Resolve dependencies :sub_project8:provided
```
The problem does **not** occur when using the `--offline` switch or when using the `--no-parallel` switch.

The problem does also occur when reducing the number of workers through the setting `org.gradle.workers.max=3`, but **not** if setting `org.gradle.workers.max=2`

The problem is reproducible and happens with every build. Apparently there is a problem with thread synchronization under certain circumstances.

### Context
This is a big multi project Gradle build with 166 sub-projects. We also use ivy repositories to resolve dependencies. I'm afraid we cannot share any gradle build files here publicly, but we can offer to share some parts privately upon request. We are happy to provide further information to help in solving this problem.

### Steps to Reproduce (for bugs)
As said above, Gradle build files cannot be attached, let alone the project's source code. And we were not able to assemble an example project. But we generated a **thread dump** of the stuck daemon process which should help to understand the problem and which is [attached](https://github.com/gradle/gradle/files/1496185/gradle.deadlock.thread.dump.txt). In this build, the number of workers was limited to 3 to narrow down the problem.

#### Our Thread analysis
*Thread "Task worker for ':' Thread 2"* holds a lock to the Gradle artifact cache here:
```
at org.gradle.cache.internal.DefaultCacheAccess.useCache(DefaultCacheAccess.java:217)
```
(locked by the `owner` member of `DefaultCacheAccess`) and waits for a different lock here:
```
at org.gradle.cache.internal.ProducerGuard$AdaptiveProducerGuard.guardByKey(ProducerGuard.java:90)
```
(trying to add a specific key to a HashSet named `producing`)

At the same time, *Thread "Daemon worker"* seems to have already added this key:
```
at org.gradle.cache.internal.ProducerGuard$AdaptiveProducerGuard.guardByKey(ProducerGuard.java:97)
```
but waits to *takeOwnership* of the Gradle artifact cache here
```
at org.gradle.cache.internal.DefaultCacheAccess.takeOwnership(DefaultCacheAccess.java:241)
```
which is held by Thread *"Task worker for ':' Thread 2"* above.

### Your Environment
The Gradle build runs on Windows (Server 2012). Relevant Gradle properties:
```
org.gradle.parallel=true
org.gradle.daemon=true
org.gradle.configureondemand=true
```
Gradle version: 4.3.1, java version: jdk1.8.0_152

### Workaround
We found a workaround that mitigates the problem by basically synchronizing all *eclipseClasspath* tasks over all workers on a task level. This works but of course slows down the build.
The code was borrowed from [another workaround](https://github.com/gradle/gradle/blob/master/gradle/fix-GRADLE-2492.gradle). We simply replaced the `generatePom` task by `eclipseClasspath` and the class by `GenerateEclipseClasspath`

**Comment by oehme** [comment]
At first glance it seems to me that DefaultArtifactResolutionQuery should not be locking the cache. @melix what do you think?

**Comment by pln-rec** [comment]
We just have seen other occurrences of the exact same problem in other tasks when Gradle needs to resolve dependencies. These are regular build tasks like `compileJava` or `test`, but the problem is not reproducible there, is rather rare. Of course, the above mentioned workaround does not fix this, so a real fix would be much appreciated. Thank you.

**Comment by blindpirate** [comment]
Step to reproduce:

- Add a line `Thread.sleep(2000)` before this line: https://github.com/gradle/gradle/blob/bce72b32e562be96a6553efe008cc26eb048ac06/subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java#L84

Use the following code can reproduce it:

```
class ArtifactResolutionQueryIntegrationTest extends AbstractHttpDependencyResolutionTest {

    @Issue('https://github.com/gradle/gradle/issues/3579')
    @Timeout(value = 60, unit = TimeUnit.SECONDS)
    def 'dead lock should not happen'() {
        given:
        mavenHttpRepo.module('group', 'artifact', '1.0').allowAll().publish()
        settingsFile << 'include "sub1","sub2"'
        buildFile << """ 


allprojects {
    apply plugin: 'java'
    repositories {
       maven { url '${mavenHttpRepo.uri}' }
    }
    dependencies {
        compile 'group:artifact:1.0'
    }
}


project(':sub1') {
  task resolveCompileSources {
      doLast {
          Thread.sleep(1000)
          def result = dependencies.createArtifactResolutionQuery()
                                   .forComponents(new org.gradle.internal.component.external.model.DefaultModuleComponentIdentifier('group','artifact','1.0'))
                                   .withArtifacts(JvmLibrary, SourcesArtifact, JavadocArtifact)
                                   .execute()
      }
    }

}
project(':sub2') {
  task retrieve(type: Sync) {
      into 'libs'
      from configurations.compile
  }
}
"""
        executer.requireOwnGradleUserHomeDir().requireIsolatedDaemons()

        expect:
        succeeds(':sub1:resolveCompileSources', 'sub2:retrieve', '--parallel')
    }
}
```

**Comment by blindpirate** [comment]
Will be fixed in `4.5RC1`.
